doc/source/data/batch_inference.rst

```diff
@@ -0,0 +1,752 @@
+
+.. caution::
+
+    Depending on how you load your data and what input data format you use, the dataset
+    loaded with :ref:`Ray Data <datasets>` will have different *batch formats*.
+    For instance, image data might be naturally stored in NumPy format, while tabular
+    data makes much more sense as a Pandas DataFrame.
+    What (default) batch format your data has and how to deal with it is explained in
+    detail in :ref:`the batch format section <batch_inference_formats>`.
+
+Defining predictors as stateful UDFs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One of the key value adds of Ray over other distributed systems is the support for
+distributed stateful operations. These stateful operations are especially useful
+for inference since the model only needs to be initialized once, instead of per batch.
+
+.. margin::
+
+    In short, running model inference means applying
+    :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`
+    to a dataset with a trained model as a UDF.
+
+You've already seen how to do this in the quickstart section of this guide, but now
+that you're equipped with more knowledge, let's have a look at how to define a
+stateful UDF with Ray for our pretrained ResNet model:
+
+.. callout::
+
+    .. literalinclude:: ./doc_code/torch_image_batch_trained.py
+        :language: python
+        :start-after: __pt_model_start__
+        :end-before: __pt_model_end__
+
+    .. annotations::
+        <1> The ``__init__`` method is used to initialize the model once. Ray takes care of distributing and managing this state for our batch processing task.
+
+        <2> The ``__call__`` method is used to apply the model to a batch of data.
+
+        <3> We're free to use any custom code in a stateful UDF, and here we prepare the data to run on GPUs.
+
+        <4> Finally, we return the ``"class"`` key of the model predictions as Numpy array.
+
+
+Scalable inference with Ray Data
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To get predictions, we call :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`,
+by making sure to specify a `ActorPoolStrategy <ray.data.compute.ActorPoolStrategy>`
+which defines how many workers to use for inference.
+
+.. callout::
+
+    .. literalinclude:: ./doc_code/torch_image_batch_trained.py
+        :language: python
+        :start-after: __pt_prediction_start__
+        :end-before: __pt_prediction_end__
+
+    .. annotations::
+        <1> In this example we use a total of four Ray Actors to run inference on our dataset.
+
+        <2> Each actor should use one GPU.
+
+To summarize, mapping a UDF over batches is the simplest transform for Ray Datastreams.
+The UDF defines the logic for transforming individual batches of data of the dataset
+Performing operations over batches of data is more performant than single element
+operations as it can leverage the underlying vectorization capabilities of Pandas or NumPy.
+
+.. _batch_inference_formats:
+
+Working with batch formats
```

Comment:
This got a lot simpler with the new `strict` mode that will be the default for Ray 2.5, we should rework this to take the changes into consideration.

@ericl can point you to the relevant changes :)

Reply:
@pcmoritz yep! I'm aware, but still wanted to have a proper "ray 2.4 checkpoint" first as a reference, as agreed with @zhe-thoughts. Will definitely overhaul this next week, once this initial version is up.